nav2_graceful_motion_controller/src/parameter_handler.cpp

```diff
@@ -0,0 +1,169 @@
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".initial_rotation", rclcpp::ParameterValue(true));
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".initial_rotation_min_angle", rclcpp::ParameterValue(0.75));
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".final_rotation", rclcpp::ParameterValue(true));
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".rotation_scaling_factor", rclcpp::ParameterValue(0.5));
+  declare_parameter_if_not_declared(
+    node, plugin_name_ + ".allow_backward", rclcpp::ParameterValue(false));
+
+  node->get_parameter(plugin_name_ + ".transform_tolerance", params_.transform_tolerance);
+  node->get_parameter(plugin_name_ + ".motion_target_dist", params_.motion_target_dist);
+  node->get_parameter(
+    plugin_name_ + ".max_robot_pose_search_dist", params_.max_robot_pose_search_dist);
+  if (params_.max_robot_pose_search_dist < 0.0) {
+    RCLCPP_WARN(
+      logger_, "Max robot search distance is negative, setting to max to search"
+      " every point on path for the closest value.");
+    params_.max_robot_pose_search_dist = std::numeric_limits<double>::max();
+  }
+
+  node->get_parameter(plugin_name_ + ".k_phi", params_.k_phi);
+  node->get_parameter(plugin_name_ + ".k_delta", params_.k_delta);
+  node->get_parameter(plugin_name_ + ".beta", params_.beta);
+  node->get_parameter(plugin_name_ + ".lambda", params_.lambda);
+  node->get_parameter(plugin_name_ + ".v_linear_min", params_.v_linear_min);
+  node->get_parameter(plugin_name_ + ".v_linear_max", params_.v_linear_max);
+  params_.v_linear_max_initial = params_.v_linear_max;
+  node->get_parameter(plugin_name_ + ".v_angular_max", params_.v_angular_max);
+  node->get_parameter(plugin_name_ + ".slowdown_radius", params_.slowdown_radius);
+  node->get_parameter(plugin_name_ + ".initial_rotation", params_.initial_rotation);
+  node->get_parameter(
+    plugin_name_ + ".initial_rotation_min_angle", params_.initial_rotation_min_angle);
+  node->get_parameter(plugin_name_ + ".final_rotation", params_.final_rotation);
+  node->get_parameter(plugin_name_ + ".rotation_scaling_factor", params_.rotation_scaling_factor);
+  node->get_parameter(plugin_name_ + ".allow_backward", params_.allow_backward);
+
+  if (params_.initial_rotation && params_.allow_backward) {
+    RCLCPP_WARN(
+      logger_, "Initial rotation and allow backward parameters are both true, "
+      "setting allow backward to false.");
+    params_.allow_backward = false;
+  }
+
+  dyn_params_handler_ = node->add_on_set_parameters_callback(
+    std::bind(&ParameterHandler::dynamicParametersCallback, this, std::placeholders::_1));
+}
+
+rcl_interfaces::msg::SetParametersResult
+ParameterHandler::dynamicParametersCallback(std::vector<rclcpp::Parameter> parameters)
+{
+  rcl_interfaces::msg::SetParametersResult result;
+  std::lock_guard<std::mutex> lock_reinit(mutex_);

```

Comment:
This lock is only in a single place. You'd need to lock the parameter `get`'s too